server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -128,6 +129,22 @@ public BytesReference internalSourceRef() {
         return XContentMapValues.extractRawValues(path, loadSourceIfNeeded());
     }
 
+    /**
+     * For each of the provided paths, return its value in the source. Note that
+     */
+    public Map<String, Object> extractValues(Collection<String> paths) {
```

Comment:
I wonder if this'd be better kept in the new fetch phase, especially if we're going to migrate away from it.

I wonder if a better interface for collecting the values is `Map<String, Consumer<?>>` or something shaped like that. I feel like it'd let you build flexible casting and formatting pretty naturally. But that can wait too.

Reply:
+1, I'll move this up into `FetchFieldsPhase`. I'll wait to refine the interface until I have a better handle on the approach to value formatting.